### PR TITLE
TR-0: Set up selenium driver

### PR DESCRIPTION
**Pull request by vadimevteev**
Zero test case, that opens testRail home page

### CODE REVIEW

File: TestProject/TestProject/UnitTest1.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+using NUnit.Framework;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Chrome;
+
+namespace TestProject;
+
+public class Tests
+{
+    private IWebDriver driver;
+    private readonly string HOME_PAGE_URL = "https://vadimevteev.testrail.io/";
```
Comment (mashadolya) [inline TestProject/TestProject/UnitTest1.cs]: Create `appsettings.json` and store all configuration data

https://weblog.west-wind.com/posts/2018/Feb/18/Accessing-Configuration-in-NET-Core-Test-Projects

File: TestProject/TestProject/UnitTest1.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        driver.Url = HOME_PAGE_URL;
+        Assert.True(driver.Url.Contains(HOME_PAGE_URL));
+    }
+
+    [TearDown]
+    public void Close()
+    {
+         driver.Quit();
+    }
+}
```
Comment (mashadolya) [inline TestProject/TestProject/UnitTest1.cs]: Please configure adding new line at the end of the file in your IDE

File: TestProject/TestProject/UnitTest1.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{
+    private IWebDriver driver;
+    private readonly string HOME_PAGE_URL = "https://vadimevteev.testrail.io/";
+    
+    [SetUp]
+    public void Start()
+    {
+        driver = new ChromeDriver();
+        driver.Manage().Window.Maximize();
+    }
```
Comment (mashadolya) [inline TestProject/TestProject/UnitTest1.cs]: Lets create `DriverFactory.cs` in order to easily access needed driver (Chrome, Firefiox etc)
And define needed driver in `appsetting.json`

**Review by mashadolya — CHANGES_REQUESTED** [review]
1. PR title must have the same name as brunch 
In your case it will be like  **TR-0: Set up selenium driver**
2. Please name the project according our test app name (TestRailAutomationTest, TestRailTest or something like that )

File: TestProject/TestProject/UnitTest1.cs (hunk 1 above)
Comment (mashadolya) [reply TestProject/TestProject/UnitTest1.cs]: Also as I remember `HOME_PAGE_URL` - is not correct C# code convention

https://github.com/ktaranov/naming-convention/blob/master/C%23%20Coding%20Standards%20and%20Naming%20Conventions.md

File: TestRailAutomationTest/Driver/DriverFactory.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace TestRailAutomationTest.Driver;
+
+public class DriverFactory
+{
+    private static IWebDriver? _driver;
+
+    public static IWebDriver GetDriver()
+    {
+        string currentBrowser = DataReader.GetConfig().Browser;
```
Comment (mashadolya) [inline TestRailAutomationTest/Driver/DriverFactory.cs]: Why not var?

File: TestRailAutomationTest/Driver/DriverFactory.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        if (_driver != null) return _driver;
```
Comment (mashadolya) [inline TestRailAutomationTest/Driver/DriverFactory.cs]: Please always put {}

File: TestRailAutomationTest/Driver/DriverFactory.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        _driver = currentBrowser switch
+        {
+            "chrome" => new ChromeDriver(),
+            "firefox" => new FirefoxDriver(),
+            _ => new EdgeDriver()
```
Comment (mashadolya) [inline TestRailAutomationTest/Driver/DriverFactory.cs]: Would be better to use chrome as default browser

File: TestRailAutomationTest/Driver/DriverFactory.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+using OpenQA.Selenium.Edge;
+using OpenQA.Selenium.Firefox;
+using TestRailAutomationTest.Service;
+
+
+namespace TestRailAutomationTest.Driver;
+
+public class DriverFactory
+{
+    private static IWebDriver? _driver;
```
Comment (mashadolya) [inline TestRailAutomationTest/Driver/DriverFactory.cs]: I think we should name it like `driver`
By the way static driver will be working only in one thread 
If we want use test parallelization we should have driver instance for each test

File: TestRailAutomationTest/Driver/DriverFactory.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    public static IWebDriver GetDriver()
+    {
+        string currentBrowser = DataReader.GetConfig().Browser;
+        if (_driver != null) return _driver;
+        _driver = currentBrowser switch
+        {
+            "chrome" => new ChromeDriver(),
+            "firefox" => new FirefoxDriver(),
```
Comment (mashadolya) [inline TestRailAutomationTest/Driver/DriverFactory.cs]: Let's create constants for this strings

File: TestRailAutomationTest/Service/DataReader.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.IO;
+using Newtonsoft.Json;
+using TestRailAutomationTest.Model;
+
+namespace TestRailAutomationTest.Service;
+
+public static class DataReader
+{
+    private static readonly string SettingsPath = @"resources\appsettings.json";
```
Comment (mashadolya) [inline TestRailAutomationTest/Service/DataReader.cs]: Should be blank line after

File: TestRailAutomationTest/Service/DataReader.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    public static Config GetConfig()
+    {
+        StreamReader stream = new StreamReader(SettingsPath);
+        string json = stream.ReadToEnd();
+        Config config = JsonConvert.DeserializeObject<Config>(json);
+        return config;
+    }
```
Comment (mashadolya) [inline TestRailAutomationTest/Service/DataReader.cs]: please always use `var`

File: TestRailAutomationTest/Service/DataReader.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+namespace TestRailAutomationTest.Service;
+
+public static class DataReader
+{
+    private static readonly string SettingsPath = @"resources\appsettings.json";
+    public static Config GetConfig()
+    {
+        StreamReader stream = new StreamReader(SettingsPath);
+        string json = stream.ReadToEnd();
+        Config config = JsonConvert.DeserializeObject<Config>(json);
```
Comment (mashadolya) [inline TestRailAutomationTest/Service/DataReader.cs]: Can be exception here 
We should handle it

File: TestRailAutomationTest/Test/HomePageTest.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+using TestRailAutomationTest.Service;
+
+namespace TestRailAutomationTest.Test;
+
+public class HomePageTest
+{
+    private IWebDriver driver;
+
+    [SetUp]
+    public void Start()
```
Comment (mashadolya) [inline TestRailAutomationTest/Test/HomePageTest.cs]: Would be better to name `Init`

**Review by mashadolya — CHANGES_REQUESTED** [review]

**Review by julsinkevich — APPROVED** [review]

**Pull request merged**